### PR TITLE
Proof of concept. Still can't throw arbitrary error...

### PR DESCRIPTION
**Pull request by Perlkonig**
DO NOT MERGE! This is proof of concept. The problem is I still can't figure out how to throw a specific error, even from the core code. Any help appreciated.
- Adds a new object to Grav\Common\Utils called [AcceptHeader](https://github.com/adoy/Accept-Header-Parser). License is embedded in the file.
- Significantly alters `Grav\Common\Twig\Twig::processSite()` as follows:
  - If a format is set either by page header or extension, it is honoured—no failsafe.
  - Otherwise, the `Accept` header is analyzed and processed in preference order.
    - The mime type `*/*` is recognized as matching `html`. This should be failsafe enough for modern user agents.
  - If a matching template is found somewhere in the search path, great. Move on.
  - Otherwise, throw an error (ideally a proper `406 Not Acceptable`).
- Right now it only supports `html`, `json`, `yaml`, and `csv` templates. Not sure if there are more that we should add.

My tests so far show this approach does work. The only think it's not doing yet is throwing the correct error. Even though I get an error message, the response code is still `200 OK`, which is not in fact OK.

Thanks for putting up with my recent code binges. I'm not a PHP native. Hopefully my coding style doesn't offend you 😃 

Related to issue #1067

### CODE REVIEW

**Comment by Perlkonig** [comment]
Of course I blew something up 😧 I'll see if I can figure out what the Travis reports mean.

**Comment by w00fz** [comment]
I think a lot of that code could be simplified if you use the `mime` method Grav offers ([see here](https://github.com/getgrav/grav/blob/develop/system/src/Grav/Common/Grav.php#L218-L247))

**Comment by w00fz** [comment]
Also for security reasons we should prevent `yaml` requests to go through. We don't want this to be exploited and potentially get access to config files, data files, password files and the likes.

**Comment by Perlkonig** [comment]
How so? The new code happens when `$format` isn't specified. The `mime` method seems to go from extension to mime type, but what I need to do is go from mime type to extension.

File: system/src/Grav/Common/Utils/AcceptHeader.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * This library is free software; you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU Lesser General Public License for more details.
+ */
```
Comment (w00fz) [inline system/src/Grav/Common/Utils/AcceptHeader.php]: I don't think we can accept this if the license is LGPL. Grav is under MIT license.

Comment (Perlkonig) [inline system/src/Grav/Common/Utils/AcceptHeader.php]: Fair enough. If that's not OK, I'll code something from scratch.

**Comment by Perlkonig** [comment]
Does .htaccess not already block direct requests for YAML files? Again, the content negotiation would only happen if there was no file extension specified in the URL. We already have the [Page as Data](https://github.com/btopro/page-as-data) plugin, for example.

**Comment by Perlkonig** [comment]
Sure. I could look at moving that logic over there. I'll do that in the next commit.

**Comment by Perlkonig** [comment]
I just realized I failed to set the `Content-Type` header, too. Next commit.

**Comment by Perlkonig** [comment]
Moved that logic into `Grav.php` and set the `templateFormat` so the correct `Content-Type` header is sent.

File: system/src/Grav/Common/Utils/AcceptHeader.php (hunk 1 above)
Comment (Perlkonig) [inline system/src/Grav/Common/Utils/AcceptHeader.php]: Latest commit removed this code. I wrote my own parser/sorter from scratch without looking 😄

File: system/src/Grav/Common/Grav.php [hunk 2]
```diff
@@ -247,6 +247,31 @@ public function mime($format)
...
+                return 'atom';
+            case 'application/rss+xml':
+                return 'rss';
+            case 'application/xml':
+                return 'xml';
+        }
+        return null;
+    }
+
+    /**
```
Comment (Perlkonig) [inline system/src/Grav/Common/Grav.php]: Done.

**Comment by flaviocopes** [comment]
To throw a specific error, use `throw new \RuntimeException('Page Not Found', 404);` for example.

**Comment by Perlkonig** [comment]
That doesn't work. If you look at the current commit, if no output is generated, I do indeed throw a `RuntimeException` with a 406 code, but all I get is an error message dump onto a blank page with a response code of `200 OK`.

**Comment by w00fz** [comment]
I might be wrong but it doesn't look like the `q` parameter for priority is doing anything. It gets set to `1.0` but on multi `Accept` headers it's always `1.0`. I see the logic in the sorter for taking `q` into account but I fail to see where this is ever going to be set differently than `1.0`.

**Comment by Perlkonig** [comment]
[The spec explains.](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) The agent provides q values. The code assigns a default q value of 1.0 if none is provided. Test it with the following header, for example:

```
application/json;q=0.5, text/html;q=0.75, application/xml
```

The correct order would be

- `application/xml`
- `text/html`
- `application/json`

**Comment by w00fz** [comment]
I see, I missed the part where you assign the original `q` back with `floatval` if it was passed in.

**Comment by w00fz** [comment]
Hey @Perlkonig I changed a bit your code, just to simplify it. I also removed some extra stuff you had that I don't think we need. The `culled` is just extra checks and loops we don't really need and doesn't really cause any issue. The sort will also take care of it grouping them at the same level. At the end of the day, you get what you need.

I removed extra details in your result array, now it's flat and only has 3 entries: `mime`, `priority` (q), `index`. Anything else wasn't used anyway.

I could just push these but I'm going to share them here first so you can see what I'm doing 😉 

```php
<?php
/**
 * Parses the `Accept` header and produces an array of acceptable
 * MIME types in preference order. If $cull is true, then only the
 * first instance of each MIME type is returned (you only care about
 * the types themselves and not the parameters).
 *
 *
 * @return mixed
 */
public static function parseAccept() {
    $result = [];

    // Break at commas
    $header = explode(',', $_SERVER['HTTP_ACCEPT']);
    
    // Look at each entry and isolate paramaters
    $result = array_map(function($entry) use ($index) {
        list($mime, $priority) = explode(';', $entry . ';q=1.0');
        $priority = explode('=', $priority)[1];
        
        return [
            'mime' => $mime,
            'q' => $priority,
            'index' => $index++
        ];
    }, $header);

    // Sort result
    usort($result, array(self, "sortAccept"));
    
    return $result;
}

/**
 * Sorts the array generated by `parseAccept` into preference order based on the spec.
 *   - q value
 *   - specificity
 *     - use of wildcards
 *     - number of parameters
 *   - order of appearance
 *
 * NOTE: Inequalities are reversed because we want the first element in the resulting
 *       array to be the one with highest priority (descending order).
 *
 * @return mixed
 */
private static function sortAccept($a, $b) {
    if ($a['q'] !== $b['q']) {
        return $a['q'] > $b['q'] ? -1 : 1;
    }

    $aAny = strpos($a['mime'], '*');
    $bAny = strpos($b['mime'], '*');
    
    if ($aAny !== $bAny) {
        return $aAny > $bAny ? -1 : 1;
    }
    
    return $a['index'] > $b['index'] ? -1 : 1;
}
```

As you can see is much simplified but does the same thing.

**Comment by Perlkonig** [comment]
TLDR: Thanks for this! I'm totally fine with pushing this to the branch.

* Thanks for cleaning things up. I'm not a PHP native, so my style is not ideal, I know.
* I did the "cull" thing because if multiple of the same mime type does exist with different parameters, then the actual content negotiation will retry the same template multiple times. Not sure what's cheaper: the cull loop or the failure of the `render()` command. Probably the latter because multiple types shouldn't frequently appear.
* I agree that most of the parameters aren't needed. I just wanted my first pass to be as inclusive as possible. I should note, though, that according to the spec and the way my sort worked, mime types aren't necessarily grouped. See last example in the spec, for instance. But again, Grav doesn't need the other parameters, so we could skip the specificity sort for now.

**Comment by rhukster** [comment]
What's the status of this PR? Is it something that is still being worked on?

**Comment by Perlkonig** [comment]
I don't know if @w00fz has more to add. I submitted it as a proof of concept. I still can't throw a specific HTTP error. If we can get it to a place where you think it will work, you should feel free to merge it, despite the perhaps-over-emphatic title. I just think it would be useful for Grav to be able to respond to a well-formed client request. As it stands, either the user has to add an extension or the page file itself has to have the response type in the front matter.

**Comment by Perlkonig** [comment]
Closing because it's old, and there's obviously no real interest. I'm going to replicate the throwing errors problem and submit that as an issue first. If that gets resolved (or is already resolved), then I'll submit a new PR against the current code base.

**Pull request closed without merging**